### DISCUSSION TITLE
Unable to add K3s cluster

### DISCUSSION THREAD

**Discussion by vvijayv** [argoproj/argo-cd#6141] [Q&A]
If you are trying to resolve an environment-specific issue or have a one-off question about the edge case that does not require a feature then please consider asking a question in argocd slack [channel](https://argoproj.github.io/community/join-slack).

Checklist:

* [x] I've searched in the docs and FAQ for my answer: https://bit.ly/argocd-faq.
* [x] I've included steps to reproduce the bug.
* [x] I've pasted the output of `argocd version`.

**Describe the bug**
I have 2 K3s in my Mac, one has the Argocd. Now, when I connect to another K3s cluster using `argocd cluster add <context-name>` I get the below error
````
INFO[0000] ServiceAccount "argocd-manager" already exists in namespace "kube-system"
INFO[0000] ClusterRole "argocd-manager-role" updated
INFO[0000] ClusterRoleBinding "argocd-manager-role-binding" updated
FATA[0001] Failed to establish connection to localhost:8080: dial tcp [::1]:8080: connect: connection refused
````

**To Reproduce**

1) create 2 k3s clusters (using k3sup or k3d)
2) Install argocd in one of the cluster following the step described in the argocd doc
3) run `argocd cluster add <2nd-cluster-context-name>`

**Expected behavior**
It should add the 2nd cluster?

**Version**

argocd: v2.0.1+33eaf11.dirty
  BuildDate: 2021-04-21T18:44:13Z
  GitCommit: 33eaf11e3abd8c761c726e815cbb4b6af7dcb030
  GitTreeState: dirty
  GoVersion: go1.16.3
  Compiler: gc
  Platform: darwin/amd64
argocd-server: v2.0.0+f5119c0
  BuildDate: 2021-04-07T06:00:33Z
  GitCommit: f5119c06686399134b3f296d44445bcdbc778d42
  GitTreeState: clean
  GoVersion: go1.16
  Compiler: gc
  Platform: linux/amd64
  Ksonnet Version: v0.13.1
  Kustomize Version: v3.9.4 2021-02-09T19:22:10Z
  Helm Version: v3.5.1+g32c2223
  Kubectl Version: v0.20.4
  Jsonnet Version: v0.17.0

```shell
INFO[0000] ServiceAccount "argocd-manager" already exists in namespace "kube-system"
INFO[0000] ClusterRole "argocd-manager-role" updated
INFO[0000] ClusterRoleBinding "argocd-manager-role-binding" updated
FATA[0001] Failed to establish connection to localhost:8080: dial tcp [::1]:8080: connect: connection refused
```

**Comment by jannfis** [comment]
You need to setup the context in your `~/.kube/config` so that it points to an URL of the K8s control plane's API reachable by the cluster Argo CD is running in. Yours is setup to connect to _localhost_, as indicated by

```
FATA[0001] Failed to establish connection to localhost:8080: dial tcp [::1]:8080: connect: connection refused
```

You may also have to reconfigure your k3s to make this work. This is not an Argo CD bug.

**Reply by vvijayv** [reply]
While I create the k3s cluster using k3d it gets added into the kube-config

**Comment by wtam2018** [comment]
@vvijayv I was able to reproduce it.  It looks like a k3d issue but we probably can't rule out `argocd cluster` command yet as kubectl is able to connect to api server.    I also tested with OpenShift cluster (which api server isn't on my localhost) and it worked fine.    You can workaround it by using the `api-port` flag to specify an address OTHER THAN localhost (127.0.0.1 or 0.0.0.0) if you have a static address or DNS address.   E.g `k3d cluster create --api-port 192.168.1.178:6443`

**Reply by vvijayv** [reply]
Let me try this

**Reply by vvijayv** [reply]
@wtam2018 I tried to have a Minikube and add the cluster using `argocd cluster add minikube` I get this error

````
$ argocd cluster add minikube                                                                                                                                    819ms  Wed May  5 06:46:23 2021
INFO[0001] ServiceAccount "argocd-manager" created in namespace "kube-system"
INFO[0001] ClusterRole "argocd-manager-role" created
INFO[0001] ClusterRoleBinding "argocd-manager-role-binding" created
FATA[0022] rpc error: code = Unavailable desc = transport is closing
````

Anything I'm missing or what may be the problem?

**Reply by wtam2018** [reply]
Have you set up an ingress?  https://argoproj.github.io/argo-cd/getting_started/#3-access-the-argo-cd-api-server  The easiest way is probably to setup port forwarding.

**Reply by vvijayv** [reply]
Yes @wtam2018, I have used the `Kube Forwarder` to expose the ArgoCD. Here is my below setup accessed in the host 
![image](https://user-images.githubusercontent.com/72122946/117247215-8154b000-ae5b-11eb-9cc6-5e2b3214ec72.png)

I tried to debug the process with `--loglevel`, but doesn't seem to work. I am not very clear on how to configure each k3d cluster with static ips.

**Reply by jgwest** [reply]
I use k3s/k3d for Argo CD development and usage. This method may not work for everyone, but  here's what I'm doing on my home network:

1) Set your machine to use a static IP (in my case I've got a dedicated machine for k3d/k3s), for example, 192.168.1.220

2) Ensure you have [ingress-nginx](https://kubernetes.github.io/ingress-nginx/) installed. ([for example](https://moonstreet.nl/post/k3s-with-ingress/)]
- Make sure it works before proceeding to the next step :smile: 

3) Create this Ingress in the Argo CD namespace
```yaml
apiVersion: extensions/v1beta1
kind: Ingress
metadata:
  name: argocd-server-ingress
  namespace: argocd
  annotations:
    kubernetes.io/ingress.class: nginx
    nginx.ingress.kubernetes.io/force-ssl-redirect: "true"
    nginx.ingress.kubernetes.io/ssl-passthrough: "true"
spec:
  rules:
  - host: cluster-argocd.192.168.1.220.nip.io
    http:
      paths:
      - backend:
          serviceName: argocd-server
          servicePort: https
```

4) `argocd login "cluster-argocd.192.168.1.220" --insecure`
